### PR TITLE
Upgrade terraform-provider-akamai to v5.0.1

### LINKED ISSUE #236: Upgrade terraform-provider-akamai to v5.0.0
Release details: https://github.com/akamai/terraform-provider-akamai/releases/tag/v5.0.0

### LINKED ISSUE #244: Upgrade terraform-provider-akamai to v5.0.1
Release details: https://github.com/akamai/terraform-provider-akamai/releases/tag/v5.0.1

### PR DESCRIPTION
**Pull request by iwahbe**
This PR was generated via `$ upgrade-provider pulumi/pulumi-akamai --major`.

---

- Updating major version from 5.0.0 to 6.0.0.
- Upgrading terraform-provider-akamai from 4.1.0  to 5.0.1.
	Fixes #244
	Fixes #236
- Upgrading pulumi-terraform-bridge from v3.53.0 to v3.54.0.
- Upgrading pulumi/terraform-plugin-sdk from v2.26.1 to v2.27.0.

## Recursive types

A large number of new files are coming from types for the underlying `akamai_imaging_policy_image` resource.

There's various different properties which might reference the root type. We identify the cycle by navigating into the $ref and comparing the list of properties and their types (ignoring the ref values as these will point to the next layer down and therefore be different).

There's also two distinct but similarly named recursion functions: `transformationTypePost` and `transformationType`. So we run the process once for each of these variations.

The process is an adaptation of [Mikhails approach from AWS to de-recurse the types](https://github.com/pulumi/pulumi-aws/pull/2589).

Before: 
image

After: 
image

Here's a list of funcitons where depth is passed down and the function calls another function with `depth - 1`:

```
  0008,15: func imageType(depth int) map[string]*schema.Schema {
  0019,25: 				Schema: boxImageType(depth - 1),
  0028,28: 				Schema: circleImageType(depth - 1),
  0037,26: 				Schema: textImageType(depth - 1),
  0046,25: 				Schema: urlImageType(depth - 1),
  0052,19: func imageTypePost(depth int) map[string]*schema.Schema {
  0063,29: 				Schema: boxImageTypePost(depth - 1),
  0072,32: 				Schema: circleImageTypePost(depth - 1),
  0081,30: 				Schema: textImageTypePost(depth - 1),
  0090,29: 				Schema: urlImageTypePost(depth - 1),
  0096,15: func shapeType(depth int) map[string]*schema.Schema {
  0107,28: 				Schema: circleShapeType(depth - 1),
  0116,27: 				Schema: pointShapeType(depth - 1),
  0125,29: 				Schema: polygonShapeType(depth - 1),
  0134,31: 				Schema: rectangleShapeType(depth - 1),
  0143,27: 				Schema: unionShapeType(depth - 1),
  0149,24: func transformationType(depth int) map[string]*schema.Schema {
  0160,23: 				Schema: appendType(depth - 1),
  0169,23: 				Schema: aspectCrop(depth - 1),
  0178,28: 				Schema: backgroundColor(depth - 1),
  0187,17: 				Schema: blur(depth - 1),
  0196,22: 				Schema: chromaKey(depth - 1),
  0205,22: 				Schema: composite(depth - 1),
  0214,21: 				Schema: compound(depth - 1),
  0223,21: 				Schema: contrast(depth - 1),
  0232,17: 				Schema: crop(depth - 1),
  0241,21: 				Schema: faceCrop(depth - 1),
  0250,24: 				Schema: featureCrop(depth - 1),
  0259,23: 				Schema: fitAndFill(depth - 1),
  0268,17: 				Schema: goop(depth - 1),
  0277,22: 				Schema: grayscale(depth - 1),
  0286,16: 				Schema: hsl(depth - 1),
  0295,16: 				Schema: hsv(depth - 1),
  0304,24: 				Schema: ifDimension(depth - 1),
  0313,26: 				Schema: ifOrientation(depth - 1),
  0322,20: 				Schema: imQuery(depth - 1),
  0331,22: 				Schema: maxColors(depth - 1),
  0340,19: 				Schema: mirror(depth - 1),
  0349,20: 				Schema: monoHue(depth - 1),
  0358,20: 				Schema: opacity(depth - 1),
  0367,33: 				Schema: regionOfInterestCrop(depth - 1),
  0376,25: 				Schema: relativeCrop(depth - 1),
  0385,24: 				Schema: removeColor(depth - 1),
  0394,19: 				Schema: resize(depth - 1),
  0403,19: 				Schema: rotate(depth - 1),
  0412,18: 				Schema: scale(depth - 1),
  0421,18: 				Schema: shear(depth - 1),
  0430,17: 				Schema: trim(depth - 1),
  0439,24: 				Schema: unsharpMask(depth - 1),
  0445,28: func transformationTypePost(depth int) map[string]*schema.Schema {
  0456,28: 				Schema: backgroundColor(depth - 1),
  0465,17: 				Schema: blur(depth - 1),
  0474,22: 				Schema: chromaKey(depth - 1),
  0483,26: 				Schema: compositePost(depth - 1),
  0492,25: 				Schema: compoundPost(depth - 1),
  0501,21: 				Schema: contrast(depth - 1),
  0510,17: 				Schema: goop(depth - 1),
  0519,22: 				Schema: grayscale(depth - 1),
  0528,16: 				Schema: hsl(depth - 1),
  0537,16: 				Schema: hsv(depth - 1),
  0546,28: 				Schema: ifDimensionPost(depth - 1),
  0555,30: 				Schema: ifOrientationPost(depth - 1),
  0564,22: 				Schema: maxColors(depth - 1),
  0573,19: 				Schema: mirror(depth - 1),
  0582,20: 				Schema: monoHue(depth - 1),
  0591,20: 				Schema: opacity(depth - 1),
  0600,24: 				Schema: removeColor(depth - 1),
  0609,24: 				Schema: unsharpMask(depth - 1),
  0616,23: func PolicyOutputImage(depth int) map[string]*schema.Schema {
  0624,24: 				Schema: breakpoints(depth - 1),
  0639,24: 				Schema: outputImage(depth - 1),
  0647,35: 				Schema: transformationTypePost(depth - 1),
  0661,31: 				Schema: transformationType(depth - 1),
  0669,21: 				Schema: variable(depth - 1),
  0675,16: func appendType(depth int) map[string]*schema.Schema {
  0704,22: 				Schema: imageType(depth - 1),
  0812,18: func boxImageType(depth int) map[string]*schema.Schema {
  0840,31: 				Schema: transformationType(depth - 1),
  0857,22: func boxImageTypePost(depth int) map[string]*schema.Schema {
  0885,35: 				Schema: transformationTypePost(depth - 1),
  0994,21: func circleImageType(depth int) map[string]*schema.Schema {
  1022,31: 				Schema: transformationType(depth - 1),
  1039,25: func circleImageTypePost(depth int) map[string]*schema.Schema {
  1067,35: 				Schema: transformationTypePost(depth - 1),
  1084,21: func circleShapeType(depth int) map[string]*schema.Schema {
  1092,27: 				Schema: pointShapeType(depth - 1),
  1109,15: func composite(depth int) map[string]*schema.Schema {
  1127,22: 				Schema: imageType(depth - 1),
  1188,19: func compositePost(depth int) map[string]*schema.Schema {
  1206,26: 				Schema: imageTypePost(depth - 1),
  1267,14: func compound(depth int) map[string]*schema.Schema {
  1273,31: 				Schema: transformationType(depth - 1),
  1279,18: func compoundPost(depth int) map[string]*schema.Schema {
  1285,35: 				Schema: transformationTypePost(depth - 1),
  1612,16: func fitAndFill(depth int) map[string]*schema.Schema {
  1619,31: 				Schema: transformationType(depth - 1),
  1788,17: func ifDimension(depth int) map[string]*schema.Schema {
  1795,31: 				Schema: transformationType(depth - 1),
  1814,31: 				Schema: transformationType(depth - 1),
  1822,31: 				Schema: transformationType(depth - 1),
  1830,31: 				Schema: transformationType(depth - 1),
  1847,21: func ifDimensionPost(depth int) map[string]*schema.Schema {
  1854,35: 				Schema: transformationTypePost(depth - 1),
  1873,35: 				Schema: transformationTypePost(depth - 1),
  1881,35: 				Schema: transformationTypePost(depth - 1),
  1889,35: 				Schema: transformationTypePost(depth - 1),
  1906,19: func ifOrientation(depth int) map[string]*schema.Schema {
  1913,31: 				Schema: transformationType(depth - 1),
  1921,31: 				Schema: transformationType(depth - 1),
  1929,31: 				Schema: transformationType(depth - 1),
  1937,31: 				Schema: transformationType(depth - 1),
  1943,23: func ifOrientationPost(depth int) map[string]*schema.Schema {
  1950,35: 				Schema: transformationTypePost(depth - 1),
  1958,35: 				Schema: transformationTypePost(depth - 1),
  1966,35: 				Schema: transformationTypePost(depth - 1),
  1974,35: 				Schema: transformationTypePost(depth - 1),
  2148,22: func polygonShapeType(depth int) map[string]*schema.Schema {
  2155,27: 				Schema: pointShapeType(depth - 1),
  2161,24: func rectangleShapeType(depth int) map[string]*schema.Schema {
  2168,27: 				Schema: pointShapeType(depth - 1),
  2196,26: func regionOfInterestCrop(depth int) map[string]*schema.Schema {
  2225,22: 				Schema: shapeType(depth - 1),
  2458,19: func textImageType(depth int) map[string]*schema.Schema {
  2517,31: 				Schema: transformationType(depth - 1),
  2533,23: func textImageTypePost(depth int) map[string]*schema.Schema {
  2592,35: 				Schema: transformationTypePost(depth - 1),
  2635,20: func unionShapeType(depth int) map[string]*schema.Schema {
  2641,22: 				Schema: shapeType(depth - 1),
  2685,18: func urlImageType(depth int) map[string]*schema.Schema {
  2692,31: 				Schema: transformationType(depth - 1),
  2708,22: func urlImageTypePost(depth int) map[string]*schema.Schema {
  2715,35: 				Schema: transformationTypePost(depth - 1),
  2731,14: func variable(depth int) map[string]*schema.Schema {
  2742,24: 				Schema: enumOptions(depth - 1),
```

### CODE REVIEW

**Comment by t0yv0** [comment]
A lot happening in 1 PR. Can you comment on / point to the changes that implement type compression?

**Comment by iwahbe** [comment]
> A lot happening in 1 PR. Can you comment on / point to the changes that implement type compression?

Type compression has not been implemented yet. So far this is just a normal update with a very large diff.

**Review by t0yv0 — APPROVED** [review]
In that case, I think this looks good to me. The update didn't los any resources, so the breaking changes seem over secondary API elements like fields on types..

File: provider/resources.go [hunk 1]
```diff
@@ -371,6 +375,90 @@ func Provider() tfbridge.ProviderInfo {
 	return prov
 }
 
+func ReplaceImagePolicyImageRecursiveTypes(spec *schema.PackageSpec, rootType string) {
```
Comment (iwahbe) [inline provider/resources.go:378]: This function doesn't actually have anything to do with `ImagePolicyImage` types. We should generalize the name.

```suggestion
func MakeRecursiveType(spec *schema.PackageSpec, rootType string) {
```

**Review by iwahbe — COMMENTED** [review]
We should add some tests, but otherwise LGTM.

I can't actually approve since I originally opened this PR.

**Review by danielrbradley — APPROVED** [review]

**Pull request merged**